substrate/pkg/controller/substrate/subnets.go

```diff
@@ -1,181 +1,137 @@
 
 import (
 	"context"
 	"fmt"
-	"strings"
 
 	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
 	"github.com/aws/aws-sdk-go/service/ec2"
-	"github.com/awslabs/kit/substrate/apis/v1alpha1"
+	"github.com/awslabs/kit/substrate/pkg/apis/v1alpha1"
+	"go.uber.org/multierr"
+	"k8s.io/client-go/util/workqueue"
 	"knative.dev/pkg/logging"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 )
 
-var (
-	azs            = []string{"us-west-2a", "us-west-2b", "us-west-2c"}
-	privateSubnets = []string{"10.0.1.0/24", "10.0.2.0/24", "10.0.3.0/24"}
-	publicSubnets  = []string{"10.0.100.0/24", "10.0.101.0/24", "10.0.102.0/24"}
-)
-
-type subnet struct {
+type subnets struct {
 	ec2Client *ec2.EC2
 }
 
-func (s *subnet) resourceName() string {
-	return "subnet"
-}
-
-func (s *subnet) Create(ctx context.Context, substrate *v1alpha1.Substrate) error {
-	if substrate.Status.VPCID == nil {
-		return fmt.Errorf("vpc ID not found for %v", substrate.Name)
+func (s *subnets) Create(ctx context.Context, substrate *v1alpha1.Substrate) (reconcile.Result, error) {
+	if substrate.Status.VPCID == nil || substrate.Status.PrivateRouteTableID == nil || substrate.Status.PublicRouteTableID == nil {
+		return reconcile.Result{Requeue: true}, nil
 	}
-	subnets, err := s.getSubnets(ctx, substrate.Name)
-	if err != nil {
-		return err
-	}
-	for i, az := range azs {
-		subnetID := ""
-		// check is private subnet exists in this AZ
-		subnet, ok := subnetExists(subnets, fmt.Sprintf("%s-private-%d", substrate.Name, i), az)
-		if !ok {
-			// create private subnet in this AZ
-			subnetID, err = s.create(ctx, substrate, az, privateSubnets[i], fmt.Sprintf("%s-private-%d", substrate.Name, i))
-			if err != nil {
-				return fmt.Errorf("provisioning private subnet, %w", err)
-			}
-		} else {
-			subnetID = *subnet.SubnetId
-		}
-		substrate.Status.PrivateSubnetIDs = append(substrate.Status.PrivateSubnetIDs, subnetID)
-
-		// check is public subnet exists in this AZ
-		subnet, ok = subnetExists(subnets, fmt.Sprintf("%s-public-%d", substrate.Name, i), az)
-		if !ok {
-			// create public subnet in this AZ
-			subnetID, err = s.create(ctx, substrate, az, publicSubnets[i], fmt.Sprintf("%s-public-%d", substrate.Name, i))
-			if err != nil {
-				return fmt.Errorf("provisioning public subnet, %w", err)
-			}
-			if err := s.markSubnetPublic(ctx, subnetID); err != nil {
-				return err
-			}
+	subnets := make([]*ec2.Subnet, len(substrate.Spec.Subnets))
+	errs := make([]error, len(substrate.Spec.Subnets))
+	workqueue.ParallelizeUntil(ctx, len(substrate.Spec.Subnets), len(substrate.Spec.Subnets), func(i int) {
+		subnets[i], errs[i] = s.ensure(ctx, substrate, substrate.Spec.Subnets[i])
+	})
+	if err := multierr.Combine(errs...); err != nil {
+		return reconcile.Result{}, err
+	}
+	for _, subnet := range subnets {
+		if aws.BoolValue(subnet.MapPublicIpOnLaunch) {
+			substrate.Status.PublicSubnetIDs = append(substrate.Status.PublicSubnetIDs, aws.StringValue(subnet.SubnetId))
```

Comment:
Should we be setting the subnetIDs in status in `ensure` method itself?

Reply:
I'd need to make the substrate threadsafe. I could do that, but felt it was more or less the same.